File: services/networkx_service/src/process/pubsub/user_publisher.py
import os
import json
from .publisher import Publisher
from network.envelope import Envelope


class UserPublisher(Publisher):
    def __init__(self, socket_id, algorithm_name, algorithm_type):
        user_metrics_topic = os.environ["USER_METRICS_TOPIC"]
        routing_key = f"{socket_id}_user_metric"
        super().__init__(user_metrics_topic, routing_key, 'direct')

        self.socket_id = socket_id
        self.algorithm_name = algorithm_name
        self.envelope = Envelope.send_user_data(self.socket_id, self.algorithm_name)

    def send(self, data):
        event_name = data['event_name']
        payload = data['payload']

        self.envelope['header']['event_name'] = event_name
        self.envelope['payload'] = payload

        super().send(json.dumps(self.envelope), 0)

File: services/jgrapht_service/src/process/src/main/java/process/pubsub/UserPublisher.java
package process.pubsub;

import org.json.JSONObject;
import process.network.Envelope;

public class UserPublisher extends Publisher {
    private final String socketId;
    private final String algorithmName;
    private final JSONObject envelope;

    public UserPublisher(String socketId, String algorithmName) {
        super(System.getenv("USER_METRICS_TOPIC"), socketId + "_user_metric", "direct");
        this.socketId = socketId;
        this.algorithmName = algorithmName;
        this.envelope = Envelope.sendUserData(socketId, algorithmName);
    }

    public void send(JSONObject data) {
        String eventName = data.getString("event_name");
        JSONObject payload = data.getJSONObject("payload");

        envelope.getJSONObject("header").put("event_name", eventName);
        envelope.put("payload", payload);

        super.send(envelope);
    }
}